ncm-chkconfig failing on CentOS 7.2

I just upgraded a machine to CentOS 7.2 and got the following error:
```
2016/01/13-09:13:22 -----------------------------------------------------------
2016/01/13-09:13:22 [ERROR] Trying to add details of unit netconsole.service but no entry in cache after adding all units and unitfiles. Ignoring this unit.
2016/01/13-09:13:25 [WARN] Found current unit netconsole.service via Chkconfig but is not in Unit. (This is unexpected, please notify the developers)
2016/01/13-09:13:48 [INFO] configure on component chkconfig executed, 1 errors, 1 warnings
```

I don't know what is `netconsole` but I am acting as requested: i'm contacting the developpers!

--- a/ncm-systemd/src/main/perl/Systemd/Service/Unit.pm
+++ b/ncm-systemd/src/main/perl/Systemd/Service/Unit.pm
@@ -721,19 +721,36 @@ sub make_cache_alias
     my @unknown;
     foreach my $unit (@units) {
         my $data = $unit_cache->{$unit};
+        my $show;
 
         if (! defined($data)) {
             my $log_method = "error";
+            my $continue;
             if (grep {$_ eq $unit} @$possible_missing) {
                 $self->debug(1, "Unit $unit is in possible_missing (and not found). ",
                              "No error will be logged.");
                 $log_method = "verbose";
+            } else {
+                $show = systemctl_show($self, $unit);
+                if(defined($show->{$PROPERTY_ID})) {
+                    $self->debug(1, "Unit $unit not listed but has show data.");
+                    $log_method = "verbose";
+                    $continue = "Found unit via systemctl show.";
+                    # Insert unit in cache
+                    $unit_cache->{$unit}->{showonly} = 1;
+                    # Reassign $data
+                    $data = $unit_cache->{$unit};
+                } else {
+                    $self->debug(1, "Unit $unit not listed but and has no show data.");
+                }
             }
+
             # no entry in cache from units or unitfiles
             $self->$log_method("Trying to add details of unit $unit ",
                          "but no entry in cache after adding all units and unitfiles. ",
-                         "Ignoring this unit.");
-            next;
+                         $continue ? $continue : "Ignoring this unit.");
+
+            next if (! $continue);
         }
 
         # check for instances
@@ -761,7 +778,7 @@ sub make_cache_alias
             }
         }
 
-        my $show = systemctl_show($self, $unit);
+        $show = systemctl_show($self, $unit) if (! defined($show));
 
         if(!defined($show)) {
             $self->error("Found unit $unit but systemctl_show returned undef. ",

--- a/ncm-systemd/src/main/perl/Systemd/Systemctl.pm
+++ b/ncm-systemd/src/main/perl/Systemd/Systemctl.pm
@@ -137,7 +137,7 @@ sub systemctl_show
     # output is k=[v]
     # some keys will be split on whitespace
     #  - when extending this list, update the pod!
-    my $res={};
+    my $res = {};
     while($output =~ m/^([^=\s]+)\s*=(.*)?$/mg) {
         my ($k,$v) = ($1,"$2");
         if (grep {$_ eq $k} @PROPERTIES_ARRAY) {

--- a/ncm-systemd/src/test/perl/service-unit.t
+++ b/ncm-systemd/src/test/perl/service-unit.t
@@ -648,6 +648,64 @@ is($cos->{'test_off.service'}->{state}, $STATE_MASKED, "test_off.service is $STA
 my $pm = $unit->possible_missing($cos);
 is_deeply($pm, [qw(test_off.service)], "Found possible missing units");
 
+=pod
 
+=head2 make_cache_alias with units that are usable but not listed
+
+=cut
+
+# set command output for list-unit / list-unit-files to EL7 output of services only
+# the netconsole.service is not listed here
+$cmdshort = "systemctl_list_unit_files_service";
+$cmdline= $cmddata::cmds{$cmdshort}{cmd};
+$out=$cmddata::cmds{$cmdshort}{out};
+$cmdline =~ s/\s+--type.*$//;
+set_desired_output($cmdline, $out);
+
+$cmdshort = "systemctl_list_units_service";
+$cmdline= $cmddata::cmds{$cmdshort}{cmd};
+$out=$cmddata::cmds{$cmdshort}{out};
+$cmdline =~ s/\s+--type.*$//;
+set_desired_output($cmdline, $out);
+
+# netconsole.service has no show output
+set_desired_output("/usr/bin/systemctl --no-pager --all show -- netconsole.service", "");
+
+# reset the cache
+$unit->init_cache();
+$cmp->{ERROR} = 0;
+
+($cache, $alias) = $unit->make_cache_alias(['netconsole.service']);
+ok(! defined($cache->{'netconsole.service'}), "no cache after unknown service (not listed, no show)");
+is($cmp->{ERROR}, 1, "an error was reported when handling unknown service (not listed, no show)");
+
+# set as possible missing
+$unit->init_cache();
+$cmp->{ERROR} = 0;
+
+($cache, $alias) = $unit->make_cache_alias(['netconsole.service'], ['netconsole.service']);
+ok(! defined($cache->{'netconsole.service'}), "no cache after unknown service (not listed, no show, possible missing)");
+is($cmp->{ERROR}, 0, "no error was reported when handling unknown service (not listed, no show, possible missing)");
+
+# minimal show data
+my $netconsole_show = <<EOF;
+Id=netconsole.service
+Names=netconsole.service
+EOF
+
+# netconsole.service has show output
+set_desired_output("/usr/bin/systemctl --no-pager --all show -- netconsole.service", $netconsole_show);
+
+# reset the cache
+$unit->init_cache();
+$cmp->{ERROR} = 0;
+
+($cache, $alias) = $unit->make_cache_alias(['netconsole.service']);
+ok($cache->{'netconsole.service'}->{showonly}, 'showonly cache attribute set when handling unknown service (not listed, with show)');
+is_deeply($cache->{'netconsole.service'}->{show}, {
+    Id => 'netconsole.service',
+    Names => ['netconsole.service'],
+}, "cache after unknown service (not listed, with show)");
+is($cmp->{ERROR}, 0, "no error was reported when handling unknown service (not listed, with show)");
 
 done_testing();

